openssl-dynamic/src/main/c/ssl.c

```diff
@@ -709,61 +979,69 @@ TCN_IMPLEMENT_CALL(jint, SSL, pendingReadableBytesInSSL)(TCN_STDARGS, jlong ssl
         return 0;
     }
-    if (w == NULL) {
+    if (wbuf == NULL) {
         tcn_ThrowException(e, "wbuf is null");
         return 0;
     }
 
     UNREFERENCED_STDARGS;
 
-    return BIO_write(b, w, wlen);
-
+    return BIO_write(bio, wbuf, wlen);
 }
 
-// Read up to rlen bytes from bio into rbuf
-TCN_IMPLEMENT_CALL(jint /* status */, SSL, readFromBIO)(TCN_STDARGS,
-                                                        jlong bio /* BIO * */,
-                                                        jlong rbuf /* char * */,
-                                                        jint rlen /* sizeof(rbuf) - 1 */) {
-    BIO *b = J2P(bio, BIO *);
-    void *r = J2P(rbuf, void *);
-
-    if (b == NULL) {
-        tcn_ThrowException(e, "bio is null");
-        return 0;
-    }
-    if (r == NULL) {
-        tcn_ThrowException(e, "rbuf is null");
-        return 0;
-    }
+TCN_IMPLEMENT_CALL(void, SSL, setBIOBuffer)(TCN_STDARGS,
+                                            jlong bioAddress /* BIO* */,
+                                            jlong bufferAddress /* Address for direct memory */,
+                                            jint maxUsableBytes /* max number of bytes to use */,
+                                            jboolean isSSLWriteSink) {
+    BIO* bio = J2P(bioAddress, BIO*);
+    char* buffer = J2P(bufferAddress, char*);
+    struct TCN_bio_bytebuffer* bioUserData = NULL;
+    TCN_ASSERT(bio != NULL);
+    TCN_ASSERT(buffer != NULL);
 
-    UNREFERENCED_STDARGS;
+    bioUserData = (struct TCN_bio_bytebuffer*) BIO_get_data(bio);
+    TCN_ASSERT(bioUserData != NULL);
 
-    return BIO_read(b, r, rlen);
+    bioUserData->buffer = buffer;
+    bioUserData->bufferLength = maxUsableBytes;
+    bioUserData->bufferIsSSLWriteSink = (bool) isSSLWriteSink;
 }
 
-TCN_IMPLEMENT_CALL(jboolean, SSL, shouldRetryBIO)(TCN_STDARGS,
-                                                        jlong bio /* BIO * */) {
-    BIO *b = J2P(bio, BIO *);
+TCN_IMPLEMENT_CALL(void, SSL, clearBIOBuffer)(TCN_STDARGS, jlong bioAddress) {
+    BIO* bio = J2P(bioAddress, BIO*);
+    struct TCN_bio_bytebuffer* bioUserData = NULL;
 
-    if (b == NULL) {
-        tcn_ThrowException(e, "bio is null");
-        return 0;
+    if (bio == NULL || (bioUserData = (struct TCN_bio_bytebuffer*) BIO_get_data(bio)) == NULL) {
+        return;
     }
 
-    UNREFERENCED_STDARGS;
+    bioUserData->buffer = NULL;
+    bioUserData->bufferLength = 0;
+    bioUserData->bufferIsSSLWriteSink = false;
+}
+
+TCN_IMPLEMENT_CALL(jint, SSL, flushBIOBuffer)(TCN_STDARGS, jlong bioAddress) {
+    BIO* bio = J2P(bioAddress, BIO*);
+    struct TCN_bio_bytebuffer* bioUserData;
 
-    return BIO_should_retry(b) ? JNI_TRUE : JNI_FALSE;
+    return (bio == NULL ||
+        (bioUserData = (struct TCN_bio_bytebuffer*) BIO_get_data(bio)) == NULL ||
+        bioUserData->sslWriteBufferLength == 0 ||
+        bioUserData->buffer == NULL ||
+        !bioUserData->bufferIsSSLWriteSink)
```

Comment:
@Scottmitch actually I think it would be easier to read if you would do the assignment etc in a separate step. I know this will cause more branches but will be easier to read at the end.